### DISCUSSION TITLE
Unable to publish the Content

### DISCUSSION THREAD

**Discussion by paritshivani** [Sunbird-Ed/Community#27] [Sunbird ED configuration discussion] [answered]
Hello Team,
Hope you are doing well!..

On Sunbird-Ed 4.4 portal setup we are facing issue while publishing content.
 Publish API is giving status 200 with message as 
"**Publish Operation for Content Id 'do_113563746228133888171' Started Successfully!**" But content is still in review state.
Publish API:
`curl 'https://devnulp.niua.org/content/content/v1/publish/do_113563746228133888171' \
  -H 'Connection: keep-alive' \
  -H 'sec-ch-ua: " Not A;Brand";v="99", "Chromium";v="98", "Google Chrome";v="98"' \
  -H 'ts: 2022-06-21T10:11:45+05:30' \
  -H 'X-msgid: c2e29643-ddd1-595e-349f-bf55bb653184' \
  -H 'X-Org-code: 0130152125009674241' \
  -H 'X-Source: web' \
  -H 'X-User-ID: 6939db9e-94ac-4748-b53f-7b2fc59b30b1' \
  -H 'sec-ch-ua-platform: "Linux"' \
  -H 'X-Channel-Id: 0130152125009674241' \
  -H 'X-App-Id: dev.dev.portal' \
  -H 'X-App-Version: 4.4.1' \
  -H 'X-Request-ID: c2e29643-ddd1-595e-349f-bf55bb653184' \
  -H 'sec-ch-ua-mobile: ?0' \
  -H 'User-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/98.0.4758.102 Safari/537.36' \
  -H 'Content-Type: application/json' \
  -H 'Accept: application/json' \
  -H 'X-Session-ID: lG8aKhCNfrdnJZ62MreuCEsH4-1nNov3' \
  -H 'X-Device-ID: c746e183f7bb3c4731ac02de4fbf7f61' \
  -H 'Origin: https://devnulp.niua.org' \
  -H 'Sec-Fetch-Site: same-origin' \
  -H 'Sec-Fetch-Mode: cors' \
  -H 'Sec-Fetch-Dest: empty' \
  -H 'Referer: https://devnulp.niua.org/workspace/content/upForReview/content/do_113563746228133888171/publish' \
  -H 'Accept-Language: en-GB,en-US;q=0.9,en;q=0.8' \
  -H 'Cookie: connect.sid=s%3AlG8aKhCNfrdnJZ62MreuCEsH4-1nNov3.6bMlDni8VN6Kb
CzjTDJxz38BMsgCK3Q4B9ZCvDI%2FQqc' \
  --data-raw '{"request":{"content":{"publishChecklist":["No Hate speech, Abuse, Violence, Profanity","No Sexual content, Nudity or Vulgarity","No Discrimination or Defamation","Is suitable for all stakeholders/ scholars","Appropriate Title, Description","Correct Category, Sub-Category, Topic, Language","Appropriate tags such as Resource Type, Concepts","Relevant Keywords","Content plays correctly","Can see the content clearly on Desktop and App","Audio (if any) is clear and easy to understand","No Spelling mistakes in the text","Language is simple to understand"],"lastPublishedBy":"6939db9e-94ac-4748-b53f-7b2fc59b30b1"}}}' \
  --compressed`

while checking logs we found some issue in knowledgemw service
we got some error as

<img width="1045" alt="Screenshot 2022-06-20 at 8 31 32 PM" src="https://user-images.githubusercontent.com/58217306/174718207-98585aca-ce75-4858-85c2-d961b0cb7387.png">

Also while checking samza logs we found following issue in logs.

`2022-06-21 06:18:30 PublishFinalizer [INFO] Ecar processing done for content: do_113563779253903360178
2022-06-21 06:18:30 PublishFinalizer [INFO] Migrating the Image Data to the Live Object. | [Content Id: do_113563779253903360178.]
2022-06-21 06:18:30 PublishFinalizer [INFO] Fetching the Content Image Node for actual state . | [Content Id: do_113563779253903360178.img]
2022-06-21 06:18:30 PublishFinalizer [INFO] Migrating the Content Body. | [Content Id: do_113563779253903360178]
2022-06-21 06:18:30 PublishFinalizer [INFO] Resetting rootNode outrelations  for content : do_113563779253903360178 as number of outrelations are :1
2022-06-21 06:18:30 PublishFinalizer [INFO] Migrating the Content Object Metadata. | [Content Id: do_113563779253903360178]
2022-06-21 06:18:30 PublishFinalizer [ERROR] Node Metadata validation failed :: ERR_GRAPH_UPDATE_NODE_VALIDATION_FAILED :: {messages=[Please select a valid framework. This framework doesn't have category: board, Please select a valid framework. This framework doesn't have category: subject, Please select a valid framework. This framework doesn't have category: medium, Please select a valid framework. This framework doesn't have category: gradeLevel]}
2022-06-21 06:18:30 PublishPipelineService [INFO] Message: Something Went Wrong While Performing 'Content Publish' Operation in Async Mode. | [Content Id: do_113563779253903360178] | event:Error! While Migrating the Metadata. | [Content Id: do_113563779253903360178]Node Metadata validation failed :: ERR_GRAPH_UPDATE_NODE_VALIDATION_FAILED :: {messages=[Please select a valid framework. This framework doesn't have category: board, Please select a valid framework. This framework doesn't have category: subject, Please select a valid framework. This framework doesn't have category: medium, Please select a valid framework. This framework doesn't have category: gradeLevel]}
2022-06-21 06:18:30 PublishPipelineService [INFO] Message: All Events sent to post publish event topic`
How can we make this working ??  

@maheshkumargangula, @amitpriyadarshi, @keshavprasadms, @pallakartheekreddy,

**Comment by paritshivani** [comment, marked as answer]
Can anyone help me here?
Issue is still unresolved

**Reply by pramodkvarma** [reply]
Error log shown above seems to have details on why it is failing (framework config issue). Did you try fixing that?

**Reply by luckynvdu** [reply]
@paritshivani please let us know if your issue is resolved with the suggested steps?

@maheshkumargangula, @amitpriyadarshi, @keshavprasadms, @pallakartheekreddy,

**Reply by maheshkumargangula** [reply]
@paritshivani - As per the logs, the framework linked to this content is not configured correctly. Please follow the below steps to identify the issue and configure the framework.

1. Use the [Content Read API](https://knowlg.sunbird.org/learn/product-and-developer-guide/content-service/content-api#content-management-apis) to get the metadata of this (do_113563746228133888171) content and identify the framework (check for the "framework" property in the content metadata).
2. Use the framework (id from the above step) and fetch the framework data using [Framework Read API](http://docs.sunbird.org/latest/apis/framework/#operation/FrameworkV1ReadGet).
3. The framework should have board, gradeLevel, medium and subject categories because they are used to tag the content.
4. Please use the [Framework Category APIs and Term APIs](https://knowlg.sunbird.org/learn/product-and-developer-guide/taxonomy-and-tagging/framework-service/apis) to configure and [publish the framework](http://docs.sunbird.org/latest/apis/framework/#operation/publishFramework).

Cc: @pramodkvarma @luckynvdu @AmiableAnil, @vinukumar-vs @amitpriyadarshi

**Reply by pranavirabatti** [reply]
Content Read API response: 
```
{
    "id": "api.content.read",
    "ver": "1.0",
    "ts": "2022-06-27T08:57:32.488Z",
    "params": {
        "resmsgid": "2e57dc80-f5f7-11ec-9195-03ede13ce066",
        "msgid": "edd11be7-d815-1fdf-06d5-17cbf864c0fc",
        "status": "successful",
        "err": null,
        "errmsg": null
    },
    "responseCode": "OK",
    "result": {
        "content": {
            "mimeType": "video/x-youtube",
            "objectType": "Content",
            "primaryCategory": "Explanation Content",
            "contentType": "Resource",
            "identifier": "do_113563746228133888171",
            "languageCode": [
                "en"
            ],
            "status": "Review",
            "contentDisposition": "online",
            "framework": "testframework",
            "createdBy": "8f424e19-962e-4372-acce-65b3c8711827",
            "resourceType": "Learn"
        }
    }
}
```
Framework Read API response:
https://pastebin.com/7V3xUmcD

@maheshkumargangula @pramodkvarma looks like everything is configured correctly. Can you please check the above two responses shared just in case we made any mistakes? Still, we are facing this issue.

**Reply by maheshkumargangula** [reply]
@pranavirabatti - Can you please restart the Content-Service and execute the API call? Looks like the in-memory cache to validate the framework data might be having an issue. Restarting the Content-Service will resolve the issue.

**Reply by pranavirabatti** [reply]
Sure will do this and let you know.
cc: @Renuka-S @komalm @coolbung

**Reply by sharathkashyap** [reply]
@pranavirabatti - Can you please confirm if the issue has been resolved?

**Reply by pranavirabatti** [reply]
@saiakhil46 can you please update here. What was the issue and what you did do to fix this?

**Reply by saiakhil46** [reply]
@pranavirabatti After following below steps, we see issue got resolved.
1. Deleted samza folder and redeployed all samza jobs
2. published the respective framework
3. And restarted the content-service
